src/config.cpp

```diff
@@ -0,0 +1,191 @@
+		return -1;
+	}
+
+	for (const auto& vendor_id : display_cfg.getMemberNames())
+	{
+		const auto& vendor = display_cfg[vendor_id];
+		if (!vendor.isObject())
+		{
+			fprintf(stderr, "Unexpected value: 'display_configuration.%s' should be an object\n",
+				vendor_id.c_str());
+			return -1;
+		}
+
+		for (const auto& product_id_str : vendor.getMemberNames())
+		{
+			char* endptr;
+			unsigned long product_id = strtoul(product_id_str.c_str(), &endptr, 0);
+			if (product_id_str.empty() || *endptr != '\0' || product_id > 0xffff)
+			{
+				fprintf(stderr, "Unexpected value: product ID '%s' in 'display_configuration.%s' should be a 16-bit integer\n",
+					product_id_str.c_str(), vendor_id.c_str());
+				return -1;
+			}
+
+			const auto& product = vendor[product_id_str];
+			if (!product.isObject())
+			{
+				fprintf(stderr, "Unexpected value: 'display_configuration.%s.%s' should be an object\n",
+					vendor_id.c_str(), product_id_str.c_str());
+				return -1;
+			}
+
+			for (const auto& name : product.getMemberNames())
+			{
+				int ret = 0;
+				if (name == "refresh_rates")
+				{
+					ret = parse_refresh_rates(product[name], vendor_id, product_id);
+				}
+				else
+				{
+					fprintf(stderr, "Unknown configuration setting 'display_configuration.%s.%s.%s'\n",
+						vendor_id.c_str(), product_id_str.c_str(), name.c_str());
+				}
+				if (ret < 0)
+					return ret;
+			}
+		}
+	}
+
+	return 0;
+}
+
+// Read a single configuration file
+static int config_read_file(const std::string& file_name)
+{
+	std::ifstream file(file_name);
+	if (!file.is_open())
+	{
+		fprintf(stderr, "Error opening config file '%s'\n", file_name.c_str());
+		return -1;
+	}
+
+	Json::Value root;
+	Json::CharReaderBuilder builder;
+	std::string errors;
+	if (!parseFromStream(builder, file, &root, &errors))
+	{
+		fprintf(stderr, "Error parsing config file '%s': %s\n",
+			file_name.c_str(), errors.c_str());
+		return -1;
+	}
+
+	if (!root.isObject())
+	{
+		fprintf(stderr, "Unexpected format in config file '%s'\n", file_name.c_str());
+		return -1;
+	}
+
+	for (const auto& name : root.getMemberNames())
+	{
+		int ret = 0;
+		if (name == "display_configuration")
+		{
+			ret = parse_display_config(root[name]);
+		}
+		else
+		{
+			fprintf(stderr, "Unknown configuration setting in '%s': '%s'\n",
+				file_name.c_str(), name.c_str());
+		}
+		if (ret < 0)
+			return ret;
+	}
+
+	return 0;
+}
+
+// Read all configuration files
+int config_read()
+{
+	std::filesystem::path config_dir = config_dir_name;
+	if (!std::filesystem::is_directory(config_dir))
+		return 0;
+
+	if (access(config_dir_name, R_OK | X_OK) != 0)
+	{
+		fprintf(stderr, "Cannot open directory '%s'\n", config_dir_name);
+		return -1;
+	}
+
+	std::vector<std::string> files;
+	for (const auto& iter : std::filesystem::directory_iterator(config_dir))
+	{
+		const auto& path = iter.path();
+		if (std::filesystem::is_regular_file(iter.status()) && path.extension() == ".conf")
```

Comment:
I'm not the fence about using a `.conf` extensions here. Since the contents are json that makes me want to lean towards the extension being `.json` instead. 

Overall, not a big deal, will go with whatever you or @Joshua-Ashton think is best

Reply:
I was actually using `.json` initially :slightly_smiling_face: but I really don't mind, both options are fine with me.